data/dataset_neo.py

```diff
@@ -0,0 +1,338 @@
+        # Add original directional edges
+        data[Constants.edge_key].edge_index = all_subgraph_edges.type(t.long)
+        data[Constants.edge_key].edge_label_index = all_sampled_edges.type(t.long)
+        data[Constants.edge_key].edge_label = labels.type(t.long)
+
+        # Add reverse edges
+        reverse_key = t.LongTensor([1, 0])
+        data[Constants.rev_edge_key].edge_index = all_subgraph_edges[reverse_key].type(
+            t.long
+        )
+        data[Constants.rev_edge_key].edge_label_index = all_sampled_edges[
+            reverse_key
+        ].type(t.long)
+        data[Constants.rev_edge_key].edge_label = labels.type(t.long)
+        return data
+
+
+def only_items_with_count_one(input: t.Tensor) -> t.Tensor:
+    uniques, counts = input.unique(return_counts=True)
+    return uniques[counts == 1]
+
+
+def get_negative_edges_random(
+    subgraph_edges_to_filter: Tensor,
+    all_edges: Tensor,
+    num_negative_edges: int,
+    randomization: bool,
+) -> Tensor:
+
+    # Get the biggest value available in articles (potential edges to sample from)
+    id_max = t.max(all_edges, dim=1)[0][1]
+
+    if all_edges.shape[1] / num_negative_edges > 100:
+        # If the number of edges is high, it is unlikely we get a positive edge, no need for expensive filter operations
+        if randomization:
+            random_integers = t.randint(
+                low=0, high=id_max.item(), size=(num_negative_edges,)
+            )
+        else:
+            random_integers = t.tensor([id_max.item()])
+
+        return random_integers
+
+    else:
+        # Create list of potential negative edges, filter out positive edges
+        only_negative_edges = only_items_with_count_one(
+            t.cat(
+                (
+                    t.arange(start=0, end=id_max + 1, dtype=t.int64),
+                    subgraph_edges_to_filter,
+                ),
+                dim=0,
+            )
+        )
+
+        # Randomly sample negative edges
+        if randomization:
+            random_integers = t.randperm(only_negative_edges.nelement())
+            negative_edges = only_negative_edges[random_integers][:num_negative_edges]
+        else:
+            negative_edges = t.tensor([id_max.item()])
+
+        return negative_edges
+
+
+def remap_edges_to_start_from_zero(
+    edges: Tensor, buckets_1st_dim: Tensor, buckets_2nd_dim: Tensor
+) -> Tensor:
+    return t.stack(
+        (
+            t.bucketize(edges[0], buckets_1st_dim),
+            t.bucketize(edges[1], buckets_2nd_dim),
+        )
+    )
+
+
+def create_edges_from_target_indices(
+    source_index: int, target_indices: Tensor
+) -> Tensor:
+    """Expand target indices list with user's id to have shape [2, num_nodes]"""
+
+    return t.stack(
+        [
+            t.Tensor([source_index]).to(dtype=t.long).repeat(len(target_indices)),
+            t.as_tensor(target_indices, dtype=t.long),
+        ],
+        dim=0,
+    )
+
+
+def fetch_n_hop_neighbourhood(
```

Comment:
should we delete this?

Reply:
I left it in so that we can compare the two results one after another. But it's just for sanity check, we can delete it otherwise.